Setting: hotkeys - ショートカットキーの設定を追加

Patchethium opened the pull request:
## Content

**1. Change the layout of setting dialog**

I'm not enthusiastic with changing UI back and forth, but when I tried to add a table in card I found it so space consuming that it's hard to show all hotkeys in a card without corrupting the visual balance. As a result, I stretched the cards a little and make the layout into a catalogue to fit the hotkey table.

![image](https://user-images.githubusercontent.com/74458240/134464362-6c113f24-957b-4767-a1a8-75d7aa0fb7e5.png)

I also changed some radios to radio buttons, just because they suits better with new layout.

**2. Add a hotkey setting table**

All the hotkey actions and combinations are shown on the card:

![image](https://user-images.githubusercontent.com/74458240/134465099-575cc2eb-9dab-4ee9-89a2-cec8e059da24.png)

Clicking one of the combinations will grey out the text and start recording keyboard combinations:

![image](https://user-images.githubusercontent.com/74458240/134465136-c44f5356-a1c0-4531-a1ad-295eee9622ec.png)

The new combination is highlighted with primary color:

![image](https://user-images.githubusercontent.com/74458240/134465177-2e3d0980-e93f-47aa-893a-0a22fbbb2375.png)

Click it again to save the combination.

![image](https://user-images.githubusercontent.com/74458240/134465199-eb279d39-00a2-46cf-a00c-442c8bf3a4e7.png)

Double click to delete the combination.

If the hotkey is also shown in menu bar, it will be updated, too.

![image](https://user-images.githubusercontent.com/74458240/134465263-741ab53a-fde6-4ba9-a9a4-e40cea3219f6.png)

![image](https://user-images.githubusercontent.com/74458240/134465250-7d4fe4b1-a07f-4201-8482-c4e0855bb31d.png)

If there's a conflict with new combos and existing combos, a dialog will show up to ask which one is to be saved.

![image](https://user-images.githubusercontent.com/74458240/134465342-22bd961c-f69d-413a-984d-c78fb5a817de.png)
(We can discuss the translation later)

There's also a search input on right corner for searching hotkeys:

![image](https://user-images.githubusercontent.com/74458240/134465446-8a16ce57-19da-48a0-86b5-ea2e0b415ad6.png)

**3. Add some new hotkey actions**

Some new hotkey actions is also added as shown below:

> 一つだけ書き出し(個別書き出し)
> 連続再生/停止
> 元に戻す
> やり直す
> テキストファイルを読み込む
> テキスト欄にフォーカスを戻す

## Involved Issues

close #255

close #253

## Others

Known bugs:

1. It's annoying to have the scroll bar above instead of below the header like in other places, but I can't figure out a way to solve it:

![image](https://user-images.githubusercontent.com/74458240/134462549-e4ed9ece-742b-4ae8-8b4d-1a9897c19567.png)

2. The setting dialog doesn't remember the last position it was.

---

## 内容

**1. 設置ダイアログのレイアウトを変えました**

あまり頻繁的にレイアウトを変えたくないんですが、ショートカットキーのテーブルには大量の空間が必要のため、カタログに移りました。

![image](https://user-images.githubusercontent.com/74458240/134464362-6c113f24-957b-4767-a1a8-75d7aa0fb7e5.png)

`radio` も`radio button`に変わりましたが、これはただカタログのスタイルに合うためです。

**2. ショートカットキーの設定を追加**

ショートカットキーとその操作が一つのテーブルに映ります

![image](https://user-images.githubusercontent.com/74458240/134465099-575cc2eb-9dab-4ee9-89a2-cec8e059da24.png)

右のショートカットをクリックすると、キーボード操作の記録が始まります

![image](https://user-images.githubusercontent.com/74458240/134465136-c44f5356-a1c0-4531-a1ad-295eee9622ec.png)

記録されたショートカットは`primary`の色でハイライトを付けます

![image](https://user-images.githubusercontent.com/74458240/134465177-2e3d0980-e93f-47aa-893a-0a22fbbb2375.png)

再びクリックして、ショートカットの登録が終わります

![image](https://user-images.githubusercontent.com/74458240/134465199-eb279d39-00a2-46cf-a00c-442c8bf3a4e7.png)

ダブルクリックしてショートカットキーを削除することもできます

メニューバーのショートカットがあれば、そこも同時に更新されます

![image](https://user-images.githubusercontent.com/74458240/134465263-741ab53a-fde6-4ba9-a9a4-e40cea3219f6.png)

![image](https://user-images.githubusercontent.com/74458240/134465250-7d4fe4b1-a07f-4201-8482-c4e0855bb31d.png)

もし重複のショートカットが指定されたら、ダイアログが出てきて、そっちを削除か、キャンセルかを聞きます

![image](https://user-images.githubusercontent.com/74458240/134465342-22bd961c-f69d-413a-984d-c78fb5a817de.png)
(日本語の翻訳ミスがありますが、マージしたあとでそっちに直すほうが早いと思います)

右上の検索欄でショートカットキーの検索もできます

![image](https://user-images.githubusercontent.com/74458240/134465446-8a16ce57-19da-48a0-86b5-ea2e0b415ad6.png)

**3. ショートカットを追加**

いくつかのショートカットでできる操作を追加しました

> 一つだけ書き出し(個別書き出し)
> 連続再生/停止
> 元に戻す
> やり直す
> テキストファイルを読み込む
> テキスト欄にフォーカスを戻す

## 関連 Issue

close #255

close #253

## その他

改善したい点

1. スクロールバーの挙動がおかしい

他のコンポーネントではヘッダーの下ですが、なぜかこっちのスクロールバーは側にあります。

![image](https://user-images.githubusercontent.com/74458240/134462549-e4ed9ece-742b-4ae8-8b4d-1a9897c19567.png)

2. この更新によって設置画面が大きくなったけど、スクロール位置が記憶されてないため、毎回位置を戻すのも手間がかかります。

Hiroshiba commented:
新しいレイアウトは一覧化されていてとてもクールです。
ショートカットキーを変更できる機能も最高にクールです。
でもこれらが１つのプルリクエストになっているのはクールじゃないです・・・！

このプルリクエストは一旦置いておいて、レイアウトを変更するだけのプルリクエストの作成をお願いします。
そちらがマージされたら、こちらに取り掛かりましょう。

レビューを待っている時間が退屈だったら、voice/unvoiceプルリクエストもお待ちしています！

---

The new layout is very cool with the listings.
The feature to change shortcut keys is also really cool.
But it's not cool that these are all in one pull request...!

Please make a pull request for only changing the layout and keep this pull request for now.
Once that one is merged, let's start working on this one.

If you' re feeling bored waiting for reviews, voice/unvoice pull requests are also welcome!

Translated with www.DeepL.com/Translator (free version)

Patchethium commented:
Okay, I created another branch exactly from 853b61d with a bit UI improvement and opened #260 . As for the voicing feature, the reason it's late is that I'm searching for a better way to get rid of that bloated `pitchHistory` design, may take some time.

Hiroshiba commented:
差分を見たいのでコンフリクト解消をよろしくお願いします！
I'd like to see the differences, so please resolve the conflicts!

Patchethium commented:
Should be okay now.

Patchethium commented on src/background.ts:
```diff
@@ -67,6 +78,76 @@ const store = new Store<{
+      },
+      {
+        action: "プロジェクトを読み込む",
+        combination: "Ctrl+O",
```
Oops, a typing mistake.

Patchethium commented:
When moving the indexes to keys I also removed the shortcuts displayed in the menu bar, they are hard to maintain and also unnecessary after we have a hotkey table.

Hiroshiba commented on src/background.ts:
```diff
@@ -44,6 +44,7 @@ protocol.registerSchemesAsPrivileged([
 const store = new Store<{
   useGpu: boolean;
   savingSetting: SavingSetting;
+  hotkeySettings: HotkeySetting[];
```
`hotkeySettings: Record<HotkeyAction, HotkeyCombo>`にするのはどうでしょう。
`for`で`HotkeyAction`を検索するコードが数箇所にあるので、楽になると思います。

---

How about making it `hotkeySettings: Record<HotkeyAction, HotkeyCombo>`?
There are a few places where the code to find `HotkeyAction` with `for`, so it would be easier.

Patchethium replied:
Record is an object so I'll have to define every hotkey not only in defaults but also in schema of the electron store. I think keeping two copies with this amount will be hard to maintain...

---

Recordはobjectですから、defaultsだけではなく、schemaにもショートカットずつ書かなければなりません。この数のコードの2つのコピーを保存するのはメンテナンス的に悪いと思います。

Hiroshiba commented on src/store/setting.ts:
```diff
@@ -1,39 +1,116 @@
   },
 } as StoreOptions<State>;
+
+export const watchHotkeys = (
+  actionKeys: HotkeyAction[],
+  hotkeyActionFunctions: (() => any)[]
+): void => {
```
`watchHotkeys`関数には２つの目的があります。１つは`HotkeyAction`に対するコールバック関数の保存、もう１つが関数の`bind/unbind`です。

関数を`bind/unbind`するのは、`SET_HOTKEY_SETTINGS`アクションに移動できます。
こうすると`store.watch`が要らなくなるのでちょっと見通しが良くなります。

もう一方の、ホットキー用のコールバック関数の保存ですが、`audioBlobCache`のようにグローバル変数を作って保存してください。
https://github.com/Hiroshiba/voicevox/blob/587bc3c5152fff994db3f34e8560bf809176009f/src/store/audio.ts#L130
これにより、`watchHotkeys`関数は`setHotkeyFunctions`になるはずです。

それと、引数の型を`[HotkeyAction, (() => any)]`か`Record<HotkeyAction, (() => any)>`にするとより見通しがよくなるはずです。

---

The `watchHotkeys` function has two purposes: one is to store callback functions for `HotkeyAction` and the other is to `bind/unbind` functions.

You can move the `bind/unbind` function to the `SET_HOTKEY_SETTINGS` action.
This way the `store.watch` will not be necessary and will be a bit more clear.

On the other hand, to store the callback functions for hotkeys, please create a global variable like `audioBlobCache`.
https://github.com/Hiroshiba/voicevox/blob/587bc3c5152fff994db3f34e8560bf809176009f/src/store/audio.ts#L130
Then, the `watchHotkeys` function will become `setHotkeyFunctions`.

Also, changing the argument type to `[HotkeyAction, (() => any)]` or `Record<HotkeyAction, (() => any)>` will make things clearer.

Hiroshiba commented on src/views/Home.vue:
```diff
@@ -236,6 +242,59 @@ export default defineComponent({
+        )
+    );
+
+    // hotkeys handled by native, for they are made to be working while focusing input elements
```
`Vue Component`のキーバインディングを動的に変更する方法がありました。
`AudioCell`内で`Dynamic Event`を使えばよりスマートに書けると思います。
https://vuejs.org/v2/guide/syntax.html#Dynamic-Arguments
https://v3.vuejs.org/api/directives.html#v-on
https://stackoverflow.com/a/48251122

---

I found a way to dynamically change the key bindings of a `Vue Component`.
Using `Dynamic Event` in `AudioCell` would be a smarter way to implement it.
https://vuejs.org/v2/guide/syntax.html#Dynamic-Arguments
https://v3.vuejs.org/api/directives.html#v-on
https://stackoverflow.com/a/48251122

Patchethium replied:
I've tried this before, somehow it turned out to be not working and I gave up. Actually the current solution **is** dynamically changing the hotkey function so there's no need to change, I think.

---

これ見たことがあるけど結局効かなかった記憶があります。実は今の取り組みでも動的に変更しているので、わざわざ書き直す必要はないと思います。

Hiroshiba reviewed (changes requested):
コードを理解しました！
ホットキーを変更するためには、ホットキーを保存し、関数をbindしたりunbindしたりする必要があるんですね。なかなか難しいタスクだ。

ただ、このままだと挙動を追うのに時間がかかってしまいます。
よりコードを見通しよくするする方法をいくつか思いつきました。

---

I understand your code now!
So in order to change a hotkey, we need to save the hotkey and bind or unbind the function. This is quite a difficult task.

But, as it is, we have to spend a lot of time tracking the behavior.
I have come up with a few ways to make the code more clear.

Patchethium commented:
It's been too much changes since the last commit and the conflicts became a mess. I'm rebuilding it from scratch and open another pr when finished.

---

当初と変わりすぎてコンフリクトがめちゃくちゃになったので立て直します。

Hiroshiba commented:
たしかにVuex付近のコードに型が入ったりして大幅に変更が生じていると思います。
プルリクエスト、とてもお待ちしております・・・！

---

Yes, I think there are significant changes in the code around Vuex caused by the addition of types.
I' m very much looking forward to your pull request...!
